myco_api_client/matching_algorithms/attributed_matching_algorithm.py

```diff
@@ -21,80 +23,95 @@ class AttributedMatchingAlgorithm(BaseMatchingAlgorithm):
 
-            if not (bids_mapping and offers_mapping):
-                continue
+                if not (bids_mapping and offers_mapping):
+                    continue
+                # Trading partners matching
+                trading_partners_recommendations = (
+                    PreferredPartnersMatchingAlgorithm.get_matches_recommendations(
+                        {market_id: data}))
+
+                bids_mapping, offers_mapping = cls._filter_out_consumed_bids_offers(
+                    bids_mapping, offers_mapping, trading_partners_recommendations)
+
+                # Green energy matching
+                green_recommendations = cls._perform_green_matching(
+                    market_id, offers_mapping, bids_mapping)
+
+                bids_mapping, offers_mapping = cls._filter_out_consumed_bids_offers(
+                    bids_mapping, offers_mapping, green_recommendations)
 
-            green_offers = cls._filter_offers_bids_by_attribute(
-                list(offers_mapping.values()), "energy_type", "PV")
-            green_bids = cls._filter_offers_bids_by_requirement(
-                list(bids_mapping.values()), "energy_type", "PV")
-            green_recommendations = PayAsBidMatchingAlgorithm.get_matches_recommendations(
-                    {market_id: {"bids": green_bids, "offers": green_offers}})
-            recommendations.extend(green_recommendations)
-            bids_mapping, offers_mapping = cls._filter_open_bids_offers(
-                bids_mapping, offers_mapping, green_recommendations)
-            residual_recommendations = PayAsBidMatchingAlgorithm.get_matches_recommendations(
-                    {market_id: {
-                        "bids": list(bids_mapping.values()),
-                        "offers": list(offers_mapping.values())}})
-            recommendations.extend(residual_recommendations)
+                # Residual matching
+                residual_recommendations = PayAsBidMatchingAlgorithm.get_matches_recommendations(
+                        {market_id: {
+                            "bids": list(bids_mapping.values()),
+                            "offers": list(offers_mapping.values())}})
+
+                recommendations.extend(
+                    trading_partners_recommendations
+                    + green_recommendations
+                    + residual_recommendations)
 
         return recommendations
 
     @classmethod
-    def _filter_offers_bids_by_requirement(
-            cls, offers_bids: List[Dict], requirement_key: str,
+    def _perform_green_matching(cls, market_id: str,
+                                offers_mapping: Dict,
+                                bids_mapping: Dict) -> List[Dict]:
+        """Check bids that require green energy and match them with valid offers."""
+        green_offers = cls._filter_orders_by_attribute(
+            list(offers_mapping.values()), "energy_type", "PV")
```

Comment:
Same here (and below): is it necessary to create a new `list`?

Reply:
Nope, this was not needed, I removed the explicit parsing

Reply:
It looks like this was not changed actually